mlflow/pyfunc/backend.py

```diff
@@ -137,33 +268,144 @@ def copy_model_into_container(dockerfile_context_dir):
+    _run_multiple_commands(
+        [
+            *activate_conda_env,
+            command,
+        ],
+        env=command_env,
+    )
+
+
+def _is_pyenv_available():
+    return _run_command(["pyenv", "--version"], check=False).returncode == 0
+
+
+def _validate_pyenv_is_available():
+    if not _is_pyenv_available():
+        raise Exception("pyenv must be available to use this feature")
+
+
+def _is_virtualenv_available():
+    return _run_command(["virtualenv", "--version"], check=False).returncode == 0
+
+
+def _validate_virtualenv_is_available():
+    if not _is_virtualenv_available():
+        raise Exception("virtualenv must be available to use this feature")
+
+
+def _get_env_name_for_virtualenv(string):
+    return "mlflow-" + hashlib.sha1(string.encode("utf-8")).hexdigest()
+
+
+def _get_mlflow_env_root():
+    return os.getenv(
+        _MLFLOW_ENV_ROOT_ENV_VAR, os.path.join(os.path.expanduser("~"), ".mlflow", "envs")
+    )
 
-        activate_conda_env += [install_mlflow]
-    if os.name != "nt":
-        separator = " && "
+
+def _install_python(version):
+    is_windows = _is_windows()
+    _logger.info("Installing python %s", version)
+    # pyenv-win doesn't support `--skip-existing` but its behavior is enabled by default
+    # https://github.com/pyenv-win/pyenv-win/pull/314
+    pyenv_install_options = () if is_windows else ("--skip-existing",)
+    _run_command(["pyenv", "install", *pyenv_install_options, version])
+
+    if is_windows:
+        # pyenv-win doesn't provide the `pyenv root` command
+        pyenv_root = os.getenv("PYENV_ROOT")
+        if pyenv_root is None:
+            raise MlflowException("Environment variable 'PYENV_ROOT' must be set")
+        path_to_bin = ("python.exe",)
     else:
-        separator = " & "
+        pyenv_root = _get_stdout(["pyenv", "root"]).strip()
+        path_to_bin = ("bin", "python")
+    return Path(pyenv_root).joinpath("versions", version, *path_to_bin)
 
-    command = separator.join(activate_conda_env + [command])
-    _logger.info("=== Running command '%s'", command)
 
-    if os.name != "nt":
-        child = subprocess.Popen(["bash", "-c", command], close_fds=True, env=command_env)
+def _execute_in_virtualenv(local_model_path, command, install_mlflow, env_id, command_env=None):
+    # Read environment information
+    model_path = Path(local_model_path)
+    python_env_file = model_path / _PYTHON_ENV_FILE_NAME
+    if python_env_file.exists():
+        python_env = PythonEnv.from_yaml(python_env_file)
     else:
-        child = subprocess.Popen(["cmd", "/c", command], close_fds=True, env=command_env)
-    rc = child.wait()
-    if rc != 0:
-        raise Exception(
-            "Command '{0}' returned non zero return code. Return code = {1}".format(command, rc)
+        _logger.info(
+            "Attempting to restore the environment using %s because this model is missing %s.",
+            _CONDA_ENV_FILE_NAME,
```

Comment:
Can we explain more specifically how we're using the `conda.yaml` file in the info message?

Reply:
Sure!